src/combinatorics/catalan.lean

```diff
@@ -0,0 +1,125 @@
+
+* `catalan n`: the `n`th Catalan number, defined recursively as
+               `catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i)`.
+* `catalan' n`: the `n`th Catalan number, defined explicitly as
+               `catalan' n = 1 / (n + 1) * nat.central_binom n`.
+
+## Main results
+
+* `catalan_defs_agree`: The fact that the two definitions both give the same sequence.
+
+## Implementation details
+
+The proof of the equality of the two definitions follows
+https://math.stackexchange.com/questions/3304415/catalan-numbers-algebraic-proof-of-the-recurrence-relation
+
+## TODO
+
+* Prove that the Catalan numbers enumerate many interesting objects.
+* Generalise to Catalan numbers associated to arbitrary Coxeter groups.
+
+-/
+
+open_locale big_operators
+open finset
+
+/-- The recursive definition of the sequence of Catalan numbers:
+`catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i)` -/
+def catalan : ℕ → ℕ
+| 0       := 1
+| (n + 1) := ∑ i : fin n.succ, have _ := i.2, have _ := nat.lt_succ_iff.mpr (n.sub_le i),
+             catalan i * catalan (n - i)
+
+@[simp] lemma catalan_zero : catalan 0 = 1 := by rw catalan
+
+lemma catalan_succ (n : ℕ) : catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i) :=
+by rw catalan
+
+/-- The explicit definition of the sequence of Catalan numbers via the central binomial
+coefficient.-/
+def catalan' (n : ℕ) :  ℚ :=
+1 / (n + 1) * nat.central_binom n
+
+@[simp] lemma catalan'_zero : catalan' 0 = 1 :=
+begin rw catalan', simp, end
+
+/-- A helper sequence that can be used to prove the equality of the recursive and the explicit
+definition using a telescoping sum argument. -/
+private def gosper_catalan (n j : ℕ) : ℚ :=
+nat.central_binom j * nat.central_binom (n - j) * (2 * j - n) / (2 * n * (n + 1))
+
+lemma gosper_trick {n i : ℕ} (h : i ≤ n) :
+gosper_catalan (n+1) (i+1) - gosper_catalan (n+1) i = catalan' i * catalan' (n - i) :=
+begin
+  have : (n:ℚ) + 1 ≠ 0 := by exact_mod_cast n.succ_ne_zero,
+  have : (n:ℚ) + 1 + 1 ≠ 0 := by exact_mod_cast (n + 1).succ_ne_zero,
+  have : (i:ℚ) + 1 ≠ 0 := by exact_mod_cast i.succ_ne_zero,
+  have : (n:ℚ) - i + 1 ≠ 0 := by exact_mod_cast (n - i).succ_ne_zero,
+  have h₁ : ((i:ℚ) + 1) * (i + 1).central_binom = 2 * (2 * i + 1) * i.central_binom,
+  { exact_mod_cast nat.succ_mul_central_binom_succ i },
+  have h₂ : ((n:ℚ) - i + 1) * (n - i + 1).central_binom
+    = 2 * (2 * (n - i) + 1) * (n - i).central_binom,
+  { exact_mod_cast nat.succ_mul_central_binom_succ (n - i) },
+  simp only [gosper_catalan, catalan'],
+  push_cast,
+  field_simp,
+  have H : n + 1 - i = n - i + 1 := sorry,
+  rw H,
+  linear_combination
+    (2:ℚ) * (n - i).central_binom * (i + 1 - (n - i)) * (n + 1) * (n + 2) * ((n - i) + 1) * h₁
+    - 2 * i.central_binom * (n + 1) * (n + 2) * (i - (n - i) - 1) * (i + 1) * h₂,
+end
+
+lemma gosper_catalan_sub_eq_catalan' (n : ℕ) :
+  gosper_catalan (n + 1) (n + 1) - gosper_catalan (n + 1) 0 = catalan' (n + 1) :=
+begin
```

Comment:
Can you do this proof by `linear_combination`?

Reply:
Sadly, `linear_combination` is not even necessary, although I'd really like to understand it better. In fact `ring` can do it. Thanks a lot for your previous code suggestion, it greatly improved that proof and in addition taught me a bit how `field_simp` is supposed to be used.